README.md
# date-expr

A Clojure library for working with date-expressions.

General functionality provided:

 * Given a date-expr pattern-string and an optional timezone,
   produce a date-expr!
 * Given a date-expr and a timestamp, compute the formatted date
 * Given a date-expr and a formatted date, parse the formatted date
   to back-compute the corresponding timestamp
 * Given a date-expr, a start time, and a stop time, compute all
   the formatted dates in between start and stop time (spaced apart by
   the finest granularity of time-period in the date expr)

Functionality around timezones:

 * Given a date-expr pattern-string, determine if it has any
   timezone conversion-specs in it
 * Given a date-expr pattern-string (that you are certain has a timezone in it)
   and a formatted date, compute the timezone of the formatted date

Functionality around shortening date-expr pattern-strings:

 * Given a date-expr pattern-string, make it "coarser" by one level of
   granularity, by removing everything before the finest-grained
   conversion-spec.
 * Given a date-expr pattern-string, make it "dateless" by removing
   everything before the first conversion-spec.

Notes:

 * Precisions are down to the second.
 * Date-expr timezone defaults to UTC if unspecified
 * Accepted timestamp formats are:
    * java.lang.Long, representing seconds since the Unix epoch
    * EpochTime, defined inside this library
    * org.joda.time.DateTime
 * Accepted timezone formats are:
    * nil (will be interpreted as utc)
    * org.joda.time.DateTimeZone
    * java.util.TimeZone
    * long form of timezone id, e.g. "America/Los_Angeles"
    * hours and minutes offset from UTC, e.g. "-0800" or "+0530"

## Terminology

 * **date-expr or date-expression**: the Clojure type representing a
   date expression
 * **date-expr pattern-string**: string that describes the format of a
   date-expr... e.g. "s3://bucket/%Y-%m-%d/foo"
 * **formatted date**: what you get when you format a date-expr using
   a particular timestamp
 * **conversion-spec or conversion specification**: the "active"
   portions of a date-expr pattern-string, which encode for time and
   get replaced when formatted. e.g. %Y is the conversion-spec for
   "four-digit year"

## Usage

Sample usage at a repl:

    > (require '[clj-time.core :as t])
    > (def ts (t/now))
    > ts
      ;; => #<DateTime 2014-08-15T15:55:15.829Z>
    > (def de (make-date-expr "s3://bucket/foo/%Y/%m/%d/bar/%H.%M/file-A"))
    > (format-expr de ts)
      ;; => "s3://bucket/foo/2014/08/15/bar/15.55/file-A"
    > (parse-expr de "s3://bucket/foo/2014/08/15/bar/15.55/file-A")
      ;; => 1408118100
    > (quot (clj-time.coerce/to-long ts) 1000)
      ;; => 1408118115 which is not quite equal to the return value of
      ;;    parse-expr... loss of precision in action!
    > (formatted-date-range (t/minus ts (t/minutes 3)) ts de)
      ;; => ("s3://bucket/foo/2014/08/15/bar/15.52/file-A"
      ;;     "s3://bucket/foo/2014/08/15/bar/15.53/file-A"
      ;;     "s3://bucket/foo/2014/08/15/bar/15.54/file-A"
      ;;     "s3://bucket/foo/2014/08/15/bar/15.55/file-A")
    > (formatted-date-range "s3://bucket/foo/2014/08/15/bar/15.52/file-A"
                            "s3://bucket/foo/2014/08/15/bar/15.55/file-A"
                            de)
      ;; => ("s3://bucket/foo/2014/08/15/bar/15.52/file-A"
      ;;     "s3://bucket/foo/2014/08/15/bar/15.53/file-A"
      ;;     "s3://bucket/foo/2014/08/15/bar/15.54/file-A"
      ;;     "s3://bucket/foo/2014/08/15/bar/15.55/file-A")
      ;; same result as the other way!

Allowable date-exprs are defined by the `date-conversion-specs` map in
the source code. It's a subset of the POSIX conversion
specifications[1]. All conversion-specs are two characters long,
beginning with a percent sign `%`. Currently (as of 2014-08-15)
escapes are not implemented, meaning that date-exprs cannot contain
the literal sequence `%Y`... that would always get interpreted to be
the year. Escapes may be implemented someday -- today is not that
day.

[1] http://linux.die.net/man/3/strftime

## License

Copyright © 2014 FIXME

Distributed under the Eclipse Public License either version 1.0 or (at
your option) any later version.


Changelog.md
# 0.3.0

 - Add docs on 'coarser' and 'dateless-prefix' (2fcd94b)
 - Add 'coarser' and 'dateless-prefix' fns (002c49c)

# 0.2.0

 - date-exprs now work with timezones (059fde1)

# 0.1.2

 - Rename danger-date-expr to date-expr (f40d0f0)

# 0.1.1

Cleanup the project.clj

 - Add url and desc to project.clj (ec22b14)
 - Fix project groupname (7c27801)

# 0.1.0

Initial public version.

 - formatted-date-range now takes formatted date-exprs (aad4991)
 - date-exprs now understand granularity AND can parse (09f54e7)
 - Define formatted-date-range (77b4a87)
 - Define core types (b4e0db2)


test/date_expr/core_test.clj
(ns date-expr.core-test
  "Tests for date-expr.core"
  {:author "Matt Halverson"
   :date "Wed Aug 13 09:48:42 PDT 2014"}
  (:require [roxxi.utils.print :refer [print-expr]])
  (:require [clojure.test :refer :all])
  (:require [date-expr.core :refer :all]
            [clj-time.core :as t])
  (:import [org.joda.time DateTimeZone]))

(def ts 1407949842) ;; 2014-08-13 at 17:10:42.000 UTC

;; # Coercion functions

(deftest coercion-test
  (let [jdt (->joda-date-time ts)
        et (->epoch-time ts)]
    (testing "Coercion functions are correct"
      (is (= (t/date-time 2014 8 13 17 10 42 0) (->joda-date-time ts)))
      (is (= (make-epoch-time 1407949842) (->epoch-time ts)))
      (is (= ts (->seconds-since-epoch ts))))

    (testing "Coercion functions are transitive (?)"
      (is (= jdt (->joda-date-time (->joda-date-time ts))))
      (is (= et (->epoch-time (->joda-date-time ts))))
      (is (= ts (->seconds-since-epoch (->joda-date-time ts))))

      (is (= jdt (->joda-date-time (->epoch-time ts))))
      (is (= et (->epoch-time (->epoch-time ts))))
      (is (= ts (->seconds-since-epoch (->epoch-time ts))))

      (testing "even if we go through another coercion step"
        (testing "(beginning with jdt)"
          (is (= jdt (->joda-date-time (->joda-date-time jdt))))
          (is (= et (->epoch-time (->joda-date-time jdt))))
          (is (= ts (->seconds-since-epoch (->joda-date-time jdt))))

          (is (= jdt (->joda-date-time (->epoch-time jdt))))
          (is (= et (->epoch-time (->epoch-time jdt))))
          (is (= ts (->seconds-since-epoch (->epoch-time jdt)))))
        (testing "(beginning with et)"
          (is (= jdt (->joda-date-time (->joda-date-time et))))
          (is (= et (->epoch-time (->joda-date-time et))))
          (is (= ts (->seconds-since-epoch (->joda-date-time et))))

          (is (= jdt (->joda-date-time (->epoch-time et))))
          (is (= et (->epoch-time (->epoch-time et))))
          (is (= ts (->seconds-since-epoch (->epoch-time et)))))))))

;; # Coarser

(deftest coarser-test
  (testing "coarser works on"
    (testing "dateless strings"
      (is (nil? (coarser "asdf"))))
    (testing "empty string"
      (is (nil? (coarser ""))))
    (testing "strings with dates at the start"
      (is (= (coarser "%Y/%m/foo/bar")
             "%Y/"))
      (is (= (coarser "%Y/foo")
             "")))
    (testing "strings with dates not at the start"
      (is (= (coarser "s3://bucket/%Y/%m/foo")
             "s3://bucket/%Y/"))
      (is (= (coarser "s3://bucket/%Y")
             "s3://bucket/"))
      (is (= (coarser "s3://bucket/%Y/%m")
             "s3://bucket/%Y/"))
      (is (= (coarser "s3://bucket/%Y/%p")
             "s3://bucket/%Y/"))
      (is (= (coarser "s3://bucket/%Y/%m/%z")
             "s3://bucket/%Y/"))
      (is (= (coarser "s3://bucket/%Y/%z")
             "s3://bucket/")))
    (testing "strings with dates that aren't in granularity order"
      (is (= (coarser "s3://bucket/%d/%m/%Y/foo")
             "s3://bucket/"))
      (is (= (coarser "%H/%d/%m/%Y/foo")
             ""))
      (is (= (coarser "%d/%H/%m/%Y/foo")
             "%d/")))))

;; # Dateless-prefix

(deftest dateless-prefix-text
  (testing "dateless-prefix works on"
    (testing "dateless strings"
      (is (= (dateless-prefix "asdf")
             "asdf")))
    (testing "empty string"
      (is (= (dateless-prefix "")
             "")))
    (testing "strings with dates at the start"
      (is (= (dateless-prefix "%Y/%m/foo/bar")
             ""))
      (is (= (dateless-prefix "%Y/foo")
             "")))
    (testing "strings with dates not at the start"
      (is (= (dateless-prefix "s3://bucket/%Y/%m/foo")
             "s3://bucket/"))
      (is (= (dateless-prefix "s3://bucket/%Y/%m")
             "s3://bucket/"))
      (is (= (dateless-prefix "s3://bucket/%Y/%p")
             "s3://bucket/"))
      (is (= (dateless-prefix "s3://bucket/%Y/%z")
             "s3://bucket/")))
    (testing "strings with dates that aren't in granularity order"
      (is (= (dateless-prefix "s3://bucket/%d/%m/%Y/foo")
             "s3://bucket/")))))

;; # Implementations of core protocol

(def date-expr (make-date-expr "s3://bucket/foo/%Y/%m/%d/bar/%H.%M/file-A"))

(deftest format-test
  (testing "We can format a DateExpr"
    (is (= (format-expr date-expr ts)
           "s3://bucket/foo/2014/08/13/bar/17.10/file-A"))
    (testing "even if it starts with a date"
      (is (= (format-expr (make-date-expr "%Y/%m/foo/bar") ts)
             "2014/08/foo/bar")))
    (testing "with a meridian specified!"
      (is (= (format-expr (make-date-expr "s3://bucket/ponies/%Y/%m/%d/%p/foo/%H.%M.%S/bar") ts)
             "s3://bucket/ponies/2014/08/13/PM/foo/17.10.42/bar")))))

(deftest parse-test
  (let [f (format-expr date-expr ts)]
    (testing "We can parse a formatted DateExpr"
      (is (= (parse-expr date-expr f)
             ;; note the loss of the "42 seconds" from ts, because date-expr
             ;; only has minute precision!
             1407949800))
      (is (= (format-expr date-expr (parse-expr date-expr f))
             f)))))

(deftest make-date-expr-test
  (testing "We can specify a TZ if we want"
    (let [de (make-date-expr "" "-0800")]
      (is (= (timezone de) (DateTimeZone/forID "-08:00")))))
  (testing "TZ defaults to utc"
    (let [de (make-date-expr "")]
      (is (= (timezone de) t/utc))))
  (testing "Malformed TZs -> exception is thrown"
    (is (thrown? Exception (make-date-expr "s3://bucket/%H.%M%z/foo" "-080")))
    (is (thrown? Exception (make-date-expr "s3://bucket/%H.%M%z/foo" "-08000")))))


;; # Timezones

(deftest has-tz-test
  (is (true? (has-tz? "%z")))
  (is (true? (has-tz? "s3://bucket/foo/%Y/%m/%d/bar/%H.%M%z/file-A")))
  (is (false? (has-tz? "s3://bucket/foo/%Y/%m/%d/bar/%H.%M%Z/file-A")))
  (is (false? (has-tz? "s3://bucket/foo/%Y/%m/%d/bar/%H.%M/file-A")))
  (is (false? (has-tz? ""))))

(deftest extract-timezone-test
  (testing "We can extract timezones from formatted strings"
    (let [s "s3://bucket/%Y/%m/%d/foo/%H.%M%z/bar"]
      (is (= (extract-timezone s "s3://bucket/2014/08/19/foo/10.43-0700/bar")
             (DateTimeZone/forID "-07:00")))
      (is (= (extract-timezone s "s3://bucket/2014/08/19/foo/10.43+0110/bar")
             (DateTimeZone/forID "+01:10")))))
  (testing "We can get back the timezone we used to format with"
    (let [des "s3://bucket/%H.%M%z/foo"
          de-1 (make-date-expr des)
          de-2 (make-date-expr des nil)
          de-3 (make-date-expr des "America/Los_Angeles")
          de-4 (make-date-expr des "-0801")
          de-5 (make-date-expr des "+0802")
          de-6 (make-date-expr des (java.util.TimeZone/getTimeZone "PST"))]
      (is (= (extract-timezone des (format-expr de-1 ts)) t/utc))
      (is (= (extract-timezone des (format-expr de-2 ts)) t/utc))
      (is (= (extract-timezone des (format-expr de-3 ts)) (DateTimeZone/forID "-07:00")))
      (is (= (extract-timezone des (format-expr de-4 ts)) (DateTimeZone/forID "-08:01")))
      (is (= (extract-timezone des (format-expr de-5 ts)) (DateTimeZone/forID "+08:02")))
      (is (= (extract-timezone des (format-expr de-6 ts)) (DateTimeZone/forID "-07:00"))))))

;; # formatted-date-range

(deftest formatted-date-range-test
  (testing "Testing the basic range-functions"
    (let [jdt (->joda-date-time ts)]
      (is (= (formatted-date-range (t/minus jdt (t/years 2)) jdt
                                   (make-date-expr "s3://bucket/foo/%Y"))
             (list
              "s3://bucket/foo/2012"
              "s3://bucket/foo/2013"
              "s3://bucket/foo/2014")))
      (is (= (formatted-date-range (t/minus jdt (t/months 4)) jdt
                                   (make-date-expr "s3://bucket/foo/%Y/%m"))
             (list
              "s3://bucket/foo/2014/04"
              "s3://bucket/foo/2014/05"
              "s3://bucket/foo/2014/06"
              "s3://bucket/foo/2014/07"
              "s3://bucket/foo/2014/08")))
      (is (= (formatted-date-range (t/minus jdt (t/weeks 1)) jdt
                                   (make-date-expr "s3://bucket/foo/%Y/%m/%d/"))
             (list
              "s3://bucket/foo/2014/08/06/"
              "s3://bucket/foo/2014/08/07/"
              "s3://bucket/foo/2014/08/08/"
              "s3://bucket/foo/2014/08/09/"
              "s3://bucket/foo/2014/08/10/"
              "s3://bucket/foo/2014/08/11/"
              "s3://bucket/foo/2014/08/12/"
              "s3://bucket/foo/2014/08/13/")))
      (is (= (formatted-date-range (t/minus jdt (t/hours 13)) jdt
                                   (make-date-expr "s3://bucket/foo/%Y/%m/%d/%p"))
             (list
              "s3://bucket/foo/2014/08/13/AM"
              "s3://bucket/foo/2014/08/13/PM")))
      (is (= (formatted-date-range (t/minus (t/minus jdt (t/hours 3)) (t/minutes 40))
                                   jdt
                                   (make-date-expr "s3://bucket/foo/%Y/%m/%d/bar/%H"))
             (list
              "s3://bucket/foo/2014/08/13/bar/13"
              "s3://bucket/foo/2014/08/13/bar/14"
              "s3://bucket/foo/2014/08/13/bar/15"
              "s3://bucket/foo/2014/08/13/bar/16")))
      (is (= (formatted-date-range (t/minus jdt (t/minutes 5)) jdt
                                   (make-date-expr "s3://bucket/foo/%Y/%m/%d/bar/%H.%M/file-A"))
             (list
              "s3://bucket/foo/2014/08/13/bar/17.05/file-A"
              "s3://bucket/foo/2014/08/13/bar/17.06/file-A"
              "s3://bucket/foo/2014/08/13/bar/17.07/file-A"
              "s3://bucket/foo/2014/08/13/bar/17.08/file-A"
              "s3://bucket/foo/2014/08/13/bar/17.09/file-A"
              "s3://bucket/foo/2014/08/13/bar/17.10/file-A")))))
  (testing "They accept args in all the combinations of types"
    (let [jdt (->joda-date-time ts)
          the-result (list
                      "s3://bucket/foo/2014/08/12/bar"
                      "s3://bucket/foo/2014/08/13/bar")
          date-expr (make-date-expr "s3://bucket/foo/%Y/%m/%d/bar")]
      (is (= (formatted-date-range (t/minus jdt (t/days 1))
                        jdt
                        date-expr)
             the-result))
      (is (= (formatted-date-range (t/minus jdt (t/days 1))
                        (->epoch-time jdt)
                        date-expr)
             the-result))
      (is (= (formatted-date-range (t/minus jdt (t/days 1))
                        (->seconds-since-epoch jdt)
                        date-expr)
             the-result))

      (is (= (formatted-date-range (->epoch-time (t/minus jdt (t/days 1)))
                        jdt
                        date-expr)
             the-result))
      (is (= (formatted-date-range (->epoch-time (t/minus jdt (t/days 1)))
                        (->epoch-time jdt)
                        date-expr)
             the-result))
      (is (= (formatted-date-range (->epoch-time (t/minus jdt (t/days 1)))
                        (->seconds-since-epoch jdt)
                        date-expr)
             the-result))

      (is (= (formatted-date-range (->seconds-since-epoch (t/minus jdt (t/days 1)))
                        jdt
                        date-expr)
             the-result))
      (is (= (formatted-date-range (->seconds-since-epoch (t/minus jdt (t/days 1)))
                        (->epoch-time jdt)
                        date-expr)
             the-result))
      (is (= (formatted-date-range (->seconds-since-epoch (t/minus jdt (t/days 1)))
                        (->seconds-since-epoch jdt)
                        date-expr)
             the-result))))
  (testing "They even accept formatted date-exprs"
    (let [s "s3://bucket/foo/2014/07/30/bar"
          f "s3://bucket/foo/2014/08/01/bar"
          date-expr (make-date-expr "s3://bucket/foo/%Y/%m/%d/bar")]
      (is (= (formatted-date-range s f date-expr)
             (list
              "s3://bucket/foo/2014/07/30/bar"
              "s3://bucket/foo/2014/07/31/bar"
              "s3://bucket/foo/2014/08/01/bar"))))))

(deftest formatted-date-range-may-use-different-timezones
  (testing (str "s and f and date-expr may all be in different timezones, "
                "but the formatted date range will be in the tz of the "
                "date-expr")
    (let [s "s3://bucket/foo/2014/08/01/bar/08.03-0800"
          f "s3://bucket/foo/2014/08/01/bar/10.06-0600"
          date-expr (make-date-expr "s3://bucket/foo/%Y/%m/%d/bar/%H.%M%z"
                                    t/utc)]
      (is (= (formatted-date-range s f date-expr)
             (list
              "s3://bucket/foo/2014/08/01/bar/16.03+0000"
              "s3://bucket/foo/2014/08/01/bar/16.04+0000"
              "s3://bucket/foo/2014/08/01/bar/16.05+0000"
              "s3://bucket/foo/2014/08/01/bar/16.06+0000"))))))


src/date_expr/core.clj
(ns date-expr.core
  "Date expr central!

  Precisions are down to the nearest second.
  Formats are in UTC timezone, unless another timezone is specified.
  Accepted time formats are:
    (1) java.lang.Long, representing seconds since the Unix epoch
    (2) EpochTime, defined inside this module
    (3) org.joda.time.DateTime"
  {:author "Matt Halverson"
   :date "Wed Aug 13 09:48:42 PDT 2014"}
  (:require [roxxi.utils.print :refer [print-expr]]
            [roxxi.utils.common :refer [def-]])
  (:require [clj-time.core :as t]
            [clj-time.periodic :as p]
            [clj-time.coerce :as c]
            [clj-time.format :as f])
  (:require [clojure.string :as s]
            [clojure.tools.logging :as log])
  (:import (org.joda.time DateTime DateTimeZone)
           (java.util TimeZone)))

;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
;; # Define an EpochTime type

(defrecord EpochTime [seconds])
(defn make-epoch-time [seconds]
  (->EpochTime seconds))


;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
;; # Coercion functions

(defmulti ->joda-date-time type)
(defmethod ->joda-date-time java.lang.Long [long]
  ;; assume that long is the number of *seconds* since epoch
  (c/from-long (* long 1000)))
(defmethod ->joda-date-time EpochTime [epoch-time]
  (->joda-date-time (:seconds epoch-time)))
(defmethod ->joda-date-time DateTime [joda-dt]
  joda-dt)
(defmethod ->joda-date-time :default [thing]
  (throw (RuntimeException.
          (format "Don't know how to convert %s to joda DateTime" (str thing)))))

(defmulti ->epoch-time type)
(defmethod ->epoch-time java.lang.Long [long]
  ;; assume that long is the number of *seconds* since epoch
  (make-epoch-time long))
(defmethod ->epoch-time EpochTime [epoch-time]
  epoch-time)
(defmethod ->epoch-time DateTime [joda-dt]
  (make-epoch-time (quot (c/to-long joda-dt) 1000)))
(defmethod ->epoch-time :default [thing]
  (throw (RuntimeException.
          (format "Don't know how to convert %s to EpochTime" (str thing)))))

(defmulti ->seconds-since-epoch type)
(defmethod ->seconds-since-epoch java.lang.Long [long]
  ;; assume that long is the number of *seconds* since epoch
  long)
(defmethod ->seconds-since-epoch EpochTime [epoch-time]
  (:seconds epoch-time))
(defmethod ->seconds-since-epoch DateTime [joda-dt]
  (quot (c/to-long joda-dt) 1000))
(defmethod ->seconds-since-epoch :default [thing]
  (throw (RuntimeException.
          (format "Don't know how to convert %s to seconds since epoch" (str thing)))))

(defn epoch-time-now []
  (->epoch-time (quot (System/currentTimeMillis) 1000)))


;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
;; # Pattern definitions (including their granularities)

(def- granularity-order
  [:year
   :month
   :day
   :meridian
   :hour
   :minute
   :second])

(defn- get-finest-granularity [granularity-seq]
  ;; XXX what if it's an empty seq?
  ;; XXX what if all are invalid granularities?
  ;; XXX what if some are invalid granularities?
  (let [positions (map #(.indexOf granularity-order %) granularity-seq)
        position-of-finest (apply max positions)]
    (get granularity-order position-of-finest)))

(def- date-conversion-specs
  ;; Joda-time patterns are defined at
  ;;  http://www.joda.org/joda-time/key_format.html
  {"%Y" {:pattern "yyyy",
         :granularity :year,
         :desc "year in four digit format"}
   "%m" {:pattern "MM",
         :granularity :month,
         :desc "month (01 to 12)"}
   "%d" {:pattern "dd",
         :granularity :day,
         :desc "day (01 to 31)"}
   "%p" {:pattern "aa",
         :granularity :meridian,
         :desc "AM or PM"}
   "%H" {:pattern "HH",
         :granularity :hour,
         :desc "hour (00 to 23)"}
   "%h" {:pattern "hh",
         :granularity :hour,
         :desc "hour (01 to 12)"}
   "%M" {:pattern "mm",
         :granularity :minute,
         :desc "minute (00 to 59)"}
   "%S" {:pattern "ss",
         :granularity :second,
         :desc "second (00 to 60 -- 60 to allow for occasional leap seconds)"}
   "%z" {:pattern "Z",
         :timezone true,
         :desc (str "The +hhmm or -hhmm numeric timezone"
                    "(that is, the hour and minute offset from UTC)"
                    "Note that there is no granularity associated with timezone."
                    "Note ALSO the :timezone key being set to true!")
         }
   })

(def- granularity=>period
  {:year (t/years 1)
   :month (t/months 1)
   :day (t/days 1)
   ;; A change-of-meridian is 11, 12, or 13 hours depending, if you allow
   ;; switching timezones to/from daylight savings halfway through. But if
   ;; you're only formatting a date-range in one timezone, like this module is,
   ;; then a change-of-meridian is always a period of 12 hours. Yay!
   :meridian (t/hours 12)
   :hour (t/hours 1)
   :minute (t/minutes 1)
   :second (t/seconds 1)})

(defn- extract-date-parts [string]
  (let [pattern-str (s/join "|" (keys date-conversion-specs))
        pattern (re-pattern pattern-str)]
    (re-seq pattern string)))

(defn- compute-granularity [string]
  (when-let [conv-specs (extract-date-parts string)]
    (let [grans (map #(get-in date-conversion-specs [% :granularity]) conv-specs)]
      (get-finest-granularity grans))))

(defn coarser
  "This computes the finest granularity of the string, then truncates
  the string up to that point.

  If the input string is void of any granularity-ful conv-specs, returns nil.

  If the input string has multiple occurrences of the finest granularity,
    truncates to the first occurrence of that granularity."
  [string]
  (when-let [g (compute-granularity string)]
    (let [matching-specs (filter #(= g (:granularity (val %))) date-conversion-specs)
          matching-strs (map first matching-specs)
          all-indices (map #(.indexOf string %) matching-strs)
          matching-indices (filter (comp not neg?) all-indices)
          first-match (apply min matching-indices)]
      (subs string 0 first-match))))

(defn- dateless? [string]
  (empty? (extract-date-parts string)))

(defn dateless-prefix
  "Given a date-expr pattern-str, returns everything in the string before
  the first conv-spec"
  [string]
  (if (dateless? string)
    string
    (let [conv-specs (extract-date-parts string)
          positions (map #(.indexOf string %) conv-specs)
          min-pos (apply min positions)]
      (subs string 0 min-pos))))

;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
;; # Core protocol

(defprotocol DateExprP
  "DateExprP is the basic, all-purpose unit of date-expression. It represents
  a date-dependent string."
  (format-expr [_ time]
    "Generates the formatted DateExpr from the given instant in time.")
  (parse-expr [_ formatted-string]
    "Extracts the approximate instant in time (seconds since the epoch) that was
    used to format a formatted DateExpr.

    Why only the approximate instant in time? Because formatting a timestamp
    may involve some loss in precision! If I format '2014/08/13 at 2 a.m.'
    to a granularity of day, I've lost the information 'at 2 a.m.'

    More generally:
      (format-expr d (parse-expr d formatted)) will get you back formatted
      (parse-expr d (format-expr d time)) will NOT get you back time")
  (timezone [_]
    "Returns the org.joda.time.DateTimeZone of this date expr"))


;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
;; # Implementation of protocol
;;
;; DateExpr knows how to operate on the types defined above
;; (EpochTime, Joda DateTime, and number of seconds since the epoch).
;;
;; It formats into UTC timezone.

(defn- empty-string? [x] (= x ""))
(defn- quote-out-pattern
  "joda DateTimeFormats will ignore parts of the pattern
  that are enclosed in single quotes. e.g. if you try to create
  a SimpleDate Format from the string s3://bucket/yyyy/MM/dd, it
  won't let you because the 'b' in bucket is an illegal pattern.

  But, it will happily work on the string 's3://bucket/'yyyy'/'MM'/'dd

  This function quotes-out the portions of the input string that
  are not date-conversion-specs."
  [string]
  (let [giant-or (s/join "|" (keys date-conversion-specs))
        re (re-pattern (str "(" giant-or ")+"))
        ;; Map first across the re-seq because the greedy matches
        ;; will return every capture group that matched... i.e.
        ;; if you have %M%z then the re-seq for that will look like
        ;; [%M%z %z]. We just want the greediest match.
        matches (map first (re-seq re string))
        splits (s/split string re)
        ;; don't replace empty string with "''"... leave it as
        ;; empty string. Why? Because "''" is how you specify an
        ;; escaped single quote! If you replace empty string
        ;; with that, you're adding a single-quote into the pattern
        ;; where there should be nothing.
        quoted-splits (map #(if (empty-string? %)
                              %
                              (str \' % \')) splits)
        interleaved (interleave quoted-splits (concat matches '("")))
        ;; (count splits) will always be one more than (count matches).
        ;; Hence, (interleave splits matches) will drop the last split!
        ;; To include the last split in there, add an empty string to
        ;; matches -- it'll disappear in the join step :)
        ]
    (s/join "" interleaved)))

;; (defn- quote-out-pattern-2
;;   [string]
;;   (let [idx-of-percents

(defn- translate-pattern [string]
  (reduce (fn [to-be-formatted kv]
            (s/replace to-be-formatted (key kv) (:pattern (val kv))))
          string
          date-conversion-specs))

(defn- ->formatter [date-expr-string ^DateTimeZone dtz]
  (let [quoted (quote-out-pattern date-expr-string)
        translated (translate-pattern quoted)]
    (f/formatter translated dtz)))

(defrecord DateExpr [string ^DateTimeZone dtz]
  DateExprP
  (format-expr [_ time]
    (let [f (->formatter string dtz)
          dt (->joda-date-time time)]
      (f/unparse f dt)))
  (parse-expr [_ formatted-string]
    (let [f (->formatter string dtz)
          dt (f/parse f formatted-string)]
      (->seconds-since-epoch dt)))
  (timezone [_]
    dtz))

(def- utc t/utc)
(defn- parse-tz [tz]
  (cond (nil? tz) (do
                    (log/debug "No timezone specified. Using UTC")
                    utc)
        (instance? DateTimeZone tz) tz
        (instance? TimeZone tz) (DateTimeZone/forTimeZone tz)
        (re-matches #"\w+/\w+" (str tz)) (DateTimeZone/forID tz)
        (re-matches #"[+|-]\d{4}" (str tz)) (DateTimeZone/forTimeZone
                                             (TimeZone/getTimeZone
                                              (str "GMT" tz)))
        :else (let [msg (format "Unrecognized timezone specified: %s" tz)]
                (log/error msg)
                (throw (RuntimeException. msg)))))

(defn make-date-expr
  ([string]
     (make-date-expr string nil))
  ([string tz]
     (->DateExpr string (parse-tz tz))))


;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
;; # Timezones

(defn- extract-tz-parts [string]
  (filter #(get-in date-conversion-specs [% :timezone])
          (extract-date-parts string)))

(defn has-tz? [string]
  (not (empty? (extract-tz-parts string))))

(defn extract-timezone
  "Given a date-expr-string and a formatted-string, returns the timezone
  of the DateExpr that was used to format the string.

  If the date-expr-string has no timezone conversion specifier, it is an error"
  [date-expr-string formatted-string]
  (when-not (has-tz? date-expr-string)
    (let [msg (format (str "Can't extract-timezone from formatted-string "
                           "(%s) because the date-expr (%s) doesn't have "
                           "a timezone in it!")
                      formatted-string
                      date-expr-string)]
      (log/error msg)
      (throw (RuntimeException. msg))))
  (let [f (->formatter date-expr-string utc)
        f' (.withOffsetParsed f)
        dt (.parseDateTime f' formatted-string)]
    (.getZone dt)))


;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
;; # Date-ranges

(defn- date-range
  "Returns a lazy seq of joda DateTimes.

  s and f represent start and finish time.
  f defaults to end of time (i.e. infinite sequence).
  period should be a joda Period."
  ([s period]
     (p/periodic-seq (->joda-date-time s) period))
  ([s f period]
     (let [s (->seconds-since-epoch s)
           f (->seconds-since-epoch f)
           ;; Round s and f to the nearest second, because this library
           ;; only provides second-level precision.
           ;;
           ;; Why? Ultimately, we round to prevent unexpected behavior in
           ;; which s is sub-second precision and f is not. i.e. if we have
           ;;    s is today at 4:00:00.556 a.m.
           ;;    f is today at 4:01:00.000 a.m.
           ;;    period-maker is t/minutes
           ;; then the seq would be (4:00:00.556) and not
           ;; (4:00:00.556, 4:01:00.556)
           ;;
           ;; So for more behavior that's (imo) more intuitive, round
           ;; to the nearest second. That means that in the above example,
           ;; the seq would be (4:00:00.000, 4:01:00.000)
           f-dt (->joda-date-time f)]
       (take-while #(or (t/before? % f-dt)
                        (= % f-dt))
                   (date-range s period)))))

(defn formatted-date-range
  "Generates a seq of dates beginning at s, separated by the finest granularity
  of the date-expr, up to the date specified by f. Returns the seq generated
  by mapping format-expr across the seq of dates.

  s and f each may be:
   - a joda DateTime
   - a java.lang.Long representing the number of seconds since the epoch
   - an EpochTime
   - a formatted date-expr.

  The type of s and the type of f may be different.

  Regarding timezones: it is allowed for s and f to be formatted in
  different timezones. A timezone is an artifact of formatting -- the
  times described by s and f are absolute. Note that, whatever the
  timezones of s and f, the formatted date-seq returned by this function
  will be in the timezone of the date-expr itself!"
  [s f date-expr]
  (let [->timestamp #(if (string? %)
                       (parse-expr date-expr %)
                       %)
        s (->timestamp s)
        f (->timestamp f)
        gran (compute-granularity (:string date-expr))
        period (get granularity=>period gran)
        the-range (date-range s f period)
        formatted-range (map #(format-expr date-expr %) the-range)]
    formatted-range))
